File: 2019/aula-16/abcpath.py
grid = None
longest = None

def dfs(y, x):
    global grid, longest
    if longest[y][x] > 0:
        return longest[y][x]
    best = 0
    for i in range(y-1, y+2):
        for j in range(x-1, x+2):
            if ord(grid[i][j]) == ord(grid[y][x])+1:
                best = max(best, dfs(i, j))
    longest[y][x]=best+1
    return longest[y][x]

def main():
    global grid, longest
    test = 0
    while True:
        height, width = [int(x) for x in input().split()]
        if width == 0 and height == 0:
            return
        grid = [['#']*(width+2)]*(height+2)
        longest = [[0]*(width+2)]*(height+2)
        for i in range(1, height+1):
            string = input()
            for j in range(1, width+1):
                grid[i][j] = string[j-1]
        best = 0
        for i in range(1, height+1):
            for j in range(1, width+1):
                if grid[i][j]=='A':
                    best = max(best, dfs(i, j))
        test+=1
        print("Case %d: %d" % (test, best))

main()

File: 2019/aula-16/abcpath.cpp
#include <cstdio>
#include <algorithm>
using namespace std;

const int maxn = 55;

char grid[maxn][maxn];
int longest[maxn][maxn];

int dfs(int y, int x) {
  if (longest[y][x] == 0) {
    int best = 0;
    for(int i=y-1; i<=y+1; ++i)
      for(int j=x-1; j<=x+1; ++j)
        if (grid[i][j] == grid[y][x]+1)
          best = max(best, dfs(i,j));
    longest[y][x]=best+1;
  }
  return longest[y][x];
}

int main() {
  int test=1;
  int height, width;
  int best;
  while(true) {
    scanf(" %d %d", &height, &width);
    if (height == 0 && width == 0)
      break;
    for(int i=0; i<=height+1; ++i)
      for(int j=0; j<=width+1; ++j){
        grid[i][j] = '#';
        longest[i][j] = 0;
      }
    for(int i=1; i<=height; ++i)
      for(int j=1; j<=width; ++j)
        scanf(" %c", &grid[i][j]);
    best=0;
    for(int i=1; i<=height; ++i)
      for(int j=1; j<=width; ++j)
        if (grid[i][j] == 'A')
          best = max(best, dfs(i,j));
    printf("Case %d: %d\n", test++, best);
  }
  return 0;
}
